File: dart_cli/lib/structs/JsonString.dart

import "../ByteDataWrapper.dart";
import "../DataType.dart";
import "../KeysLookup.dart";
import "../SizeClass.dart";
import "../ioClass.dart";

const _sizeClassToType = {
  SizeClass.empty: DataTypeChar.emptyString,
  SizeClass.small: DataTypeChar.smallString,
  SizeClass.big: DataTypeChar.bigString,
  SizeClass.long: DataTypeChar.longString
};
const _typeToSizeClass = {
  DataTypeChar.emptyString: SizeClass.empty,
  DataTypeChar.smallString: SizeClass.small,
  DataTypeChar.bigString: SizeClass.big,
  DataTypeChar.longString: SizeClass.long
};

class JsonString extends JsonIoClass {
  final int length;
  final String value;
  final SizeClass sizeClass;

  JsonString(this.length, this.value, this.sizeClass);

  static (DataType, JsonString?) fromJson(Object? json) {
    if (json is! String)
      throw Exception("JsonString.fromJson: json is not a String: $json");
    var string = json;
    var length = ByteDataWrapper.encodeString(string).length;
    var sizeClass = SizeClass.fromSize(length);
    var type = _sizeClassToType[sizeClass];
    if (type == null)
      throw Exception("JsonString.fromJson: invalid size class: $sizeClass");
    return (
      DataType(type),
      sizeClass != SizeClass.empty
        ? JsonString(length, string, sizeClass)
        : null
    );
  }

  static JsonString readBytes(ByteDataWrapper bytes, DataType dataType) {
    var sizeClass = _typeToSizeClass[dataType.char];
    if (sizeClass == null)
      throw Exception("JsonString.readBytes: dataType is not a string: $dataType");
    var length = sizeClass.readInt(bytes);
    var value = bytes.readString(length);
    return JsonString(length, value, sizeClass);
  }

  @override
  Object? toJson() {
    return value;
  }

  @override
  void writeBytes(ByteDataWrapper bytes, KeysLookup keysLookup) {
    sizeClass.writeInt(bytes, length);
    bytes.writeString(value);
  }

  @override
  int get size {
    return sizeClass.bytes + length;
  }
}

File: python_lib/structs/JsonString.py
from __future__ import annotations
from DataType import DataType, DataTypeChar
from FileWrapper import BinaryReader, BinaryWriter
from KeysLookup import KeysLookup
from SizeClass import SizeClass
from ioClass import JsonIoClass

_sizeClassToType = {
	SizeClass.empty: DataTypeChar.emptyString,
	SizeClass.small: DataTypeChar.smallString,
	SizeClass.big: DataTypeChar.bigString,
	SizeClass.long: DataTypeChar.longString
}
_typeToSizeClass = {
	DataTypeChar.emptyString: SizeClass.empty,
	DataTypeChar.smallString: SizeClass.small,
	DataTypeChar.bigString: SizeClass.big,
	DataTypeChar.longString: SizeClass.long
}

class JsonString(JsonIoClass):
	length: int
	value: str
	sizeClass: SizeClass

	def __init__(self, length: int, value: str, sizeClass: SizeClass):
		self.length = length
		self.value = value
		self.sizeClass = sizeClass

	@staticmethod
	def fromJson(json: object) -> tuple[DataType, JsonString|None]:
		if not isinstance(json, str):
			raise Exception(f"JsonString.fromJson: json is not a String: {json}")
		string = json
		length = len(string.encode())
		sizeClass = SizeClass.fromSize(length)
		type = _sizeClassToType[sizeClass]
		if type is None:
			raise Exception(f"JsonString.fromJson: invalid size class: {sizeClass}")
		return (
			DataType(type),
			JsonString(length, string, sizeClass) if sizeClass is not SizeClass.empty else None
		)
	
	@staticmethod
	def readBytes(bytes: BinaryReader, dataType: DataType) -> JsonString:
		sizeClass = _typeToSizeClass[dataType.char]
		if sizeClass is None:
			raise Exception(f"JsonString.readBytes: dataType is not a string: {dataType}")
		length = sizeClass.readInt(bytes)
		value = bytes.readString(length)
		return JsonString(length, value, sizeClass)
	
	def toJson(self) -> object:
		return self.value
	
	def writeBytes(self, bytes: BinaryWriter, keysLookup: KeysLookup) -> None:
		self.sizeClass.writeInt(bytes, self.length)
		bytes.writeString(self.value)
	
	@property
	def size(self) -> int:
		return self.sizeClass.bytes + self.length
